### PR TITLE
ARROW-6778: [C++] Support cast for DurationType

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
https://issues.apache.org/jira/browse/ARROW-6778

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/cast.cc [hunk 1]
```diff
@@ -477,13 +477,19 @@ const std::pair<bool, int64_t> kTimeConversionTable[4][4] = {
...
-struct CastFunctor<TimestampType, TimestampType> {
+// <TimestampType, TimestampType> and <DurationType, DurationType>
+template <typename O, typename I>
+struct CastFunctor<
+    O, I,
+    typename std::enable_if<(std::is_base_of<O, TimestampType>::value &&
+                             std::is_base_of<I, TimestampType>::value) ||
+                            (std::is_base_of<O, DurationType>::value &&
+                             std::is_base_of<I, DurationType>::value)>::type> {
   void operator()(FunctionContext* ctx, const CastOptions& options,
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/kernels/cast.cc]: So for some reason, this works for `<TimestampType, TimestampType>`, but not for `<DurationType, DurationType>`. 
This functions calls `ShiftTime<int64_t, int64_t>` (defined just above), and in there, it is the line

```
auto out_data = output->GetMutableValues<out_type>(1);
```

that segfaults. But I would think that TimestampType and DurationType have the exact same layout, and should not behave differently?

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: Is that line crashes, then it probably means `output` is null.

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: So, the issue is probably that `is_primitive` (in `arrow/type_traits.h`) returns false for Duration. It should probably be fixed.

Comment (jorisvandenbossche) [reply cpp/src/arrow/compute/kernels/cast.cc]: Thanks, that indeed seems to fix it

**Comment by jorisvandenbossche** [comment]
Still to do in this PR is casting to different unit.

**Comment by pitrou** [comment]
> Still to do in this PR is casting to different unit.

What do you mean? It seems this is tested for.

**Comment by jorisvandenbossche** [comment]
Ah, yes, you're right ;) I somehow forgot I already added that ... 
It was already implemented, and tested for going to higher resolution (s -> ms), but not yet for the other way around, so added tests for that now.

**Pull request closed without merging**